README
This is Perl module Net::Jenkins.

INSTALLATION

Net::Jenkins installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Net::Jenkins

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Net::Jenkins documentation is available as in POD. So you can do:

    % perldoc Net::Jenkins

to read the documentation online with your favorite pager.

Yo-An Lin


Changes
Revision history for Perl extension Net::Jenkins

0.06    六  7/21 10:04:25 2012
        - Add with_details flag

0.05    六  5/26 12:09:32 2012
        - Add to_hashref method

0.04    六  5/26 11:17:36 2012
        - Update pod documentation
        - Rename _api to api

0.03    五  5/25 01:03:57 2012
        - Add Net::Jenkins::Utils

0.02    三  5/23 10:38:39 2012
        - Fix dependency

0.01    Wed May 23 00:15:26 2012
        - original version


lib/Net/Jenkins/Job.pm
package Net::Jenkins::Job;
use methods;
use Moose;
use Net::Jenkins::Job::Build;
use Net::Jenkins::Job::QueueItem;

has name => ( is => 'rw' , isa => 'Str' );

# 'color' => 'grey',
has color => ( is => 'rw' );

# 'url' => 'http://localhost:8080/job/Phifty/',
has url => ( is => 'rw' , isa => 'Str' );

has api => ( is => 'rw' , isa => 'Net::Jenkins' );

method delete {
    return $self->api->delete_job( $self->name );
}

method update ($xml) {
    return $self->api->update_job( $self->name, $xml );
}

method copy ($new_job_name) {
    return $self->api->copy_job( $new_job_name , $self->name );
}


method enable {
    return $self->api->enable_job($self->name);
}

method disable {
    return $self->api->disable_job($self->name);
}


# trigger a build
method build {
    return $self->api->build_job($self->name);
}


# get job configuration
method details {
    return $self->api->get_job_details( $self->name );
}

method description {
    return $self->details->{description};
}

method desc {
    return $self->description;
}

method in_queue {
    return $self->details->{inQueue};
}

method queue_item {
    return Net::Jenkins::Job::QueueItem->new( %{ $self->details->{queueItem} } , api => $self->api , job => $self );
}

# get builds
method builds {
    return map { Net::Jenkins::Job::Build->new( %$_ , api => $self->api, job => $self ) } 
            $self->api->get_builds( $self->name );
}

method last_build {
    my $b = $self->details->{lastBuild};
    return Net::Jenkins::Job::Build->new( %$b , api => $self->api , job => $self ) if $b && %$b;
}

method first_build {
    my $b = $self->details->{firstBuild};
    return Net::Jenkins::Job::Build->new( %$b , api => $self->api , job => $self ) if $b && %$b;
}

method to_hashref ($with_details) {
    return {
        name    => $self->name,
        color   => $self->color,
        url     => $self->url,
        ($with_details) ? (
            details => $self->details,
            builds => [ map { 
                    $_->to_hashref($with_details) 
                } $self->builds ],
        ) : ()
    };
}

1;
__END__
=pod

=head1 NAME

Net::Jenkins::Job

=head1 ATTRIBUTES

=head2 name

=head2 color

=head2 url

=head2 api

=head1 METHODS

=head2 delete

Delete this job.

=head2 update ($xml)

$xml [Str]

Update job configuration from XML.

=head2 copy ($new_job_name)

$new_job_name [Str]

Copy from a job.

=head2 enable 

Enable this job.

=head2 disable

Disable this job.

=head2 build

Trigger build

=head2 details

[HashRef] Get job details

=head2 description

[Str] Get job description

=head2 in_queue

[Bool] Is this job in queue ?

=head2 builds

L<Net::Jenkins::Job::Build>[] get build objects.

=head2 last_build

L<Net::Jenkins::Job::Build> Get last build.

=head2 first_build

L<Net::Jenkins::Job::Build> Get first build.

=cut


lib/Net/Jenkins/Utils.pm
package Net::Jenkins::Utils;
use warnings;
use strict;
use URI;
use Net::Jenkins;
use Net::Jenkins::Job;
use Net::Jenkins::Job::Build;
use parent 'Exporter';

our @EXPORT_OK = qw(build_job_object build_build_object build_api_object);

sub build_job_object;
sub build_api_object;
sub build_build_object;

sub build_api_object ($) {
    my $url = shift;
    my $uri = ref($url) eq 'URI' ? $url : URI->new( $url );
    return Net::Jenkins->new( 
        host => $uri->host, 
        port => $uri->port , 
        scheme => $uri->scheme );
}

sub build_build_object ($) {
    my $build_url = shift;
    my $uri = URI->new( $build_url );

    # http://ci.jruby.org/job/jruby-git/4259/api/json
    my ($job_name,$build_id) = 
        ($build_url =~ m{/job/([^/]+)/([^/]+)});


    my ($job_url) = ($build_url =~ m{^(.*/job/[^/]+)});

    my $job = build_job_object $job_url;
    my $build = Net::Jenkins::Job::Build->new(
        number => $build_id,
        url => $build_url,
        job => $job,
        _api => build_api_object($job_url),
    );
    return $build;
}

sub build_job_object ($) {
    my $job_url = shift;
    my $uri = URI->new($job_url);
    my ($job_name) = ($job_url =~ m{job/([^/]+)});

    my $job = Net::Jenkins::Job->new( 

        # XXX: color attribute
        name => $job_name,
        url => $job_url,
        _api => build_api_object $job_url,
    );
    return $job;
}


1;
__END__

=head2 build_job_object (Str $job_url)

@return Net::Jenkins::Job

=cut




lib/Net/Jenkins/Job/Build.pm
package Net::Jenkins::Job::Build;
use Moose;
use methods;
use DateTime;


# {
# 'number' => 2,
# 'url' => 'http://localhost:8080/job/Phifty/2/'
# },

has number => ( is => 'rw' , isa => 'Int' );

has url => ( is => 'rw', isa => 'Str' );

has job => ( is => 'rw', isa => 'Net::Jenkins::Job' );

has api => ( is => 'rw' , isa => 'Net::Jenkins' );

method details {
    return $self->api->get_build_details( $self->job->name, $self->number );
}

method result {
    return $self->details->{result};
}

method building {
    return $self->details->{building};
}

method id {
    return $self->details->{id};
}

method name {
    return $self->details->{fullDisplayName};
}

method created_at {
    return DateTime->from_epoch( epoch => $self->details->{timestamp} );
}

method console {
    return $self->api->get_build_console( $self->job->name, $self->number );
}

method console_handle {
    return $self->api->get_build_console_handle( $self->job->name , $self->number );
}

method to_hashref {
    return {
        number => $self->number,
        url => $self->url,
        job => $self->job->to_hashref,
        details => $self->details,
    };
}

1;
__END__
=head1 NAME

Net::Jenkins::Job::Build

=head1 ATTRIBUTES

=head2 number

[Int] The build number

=head2 url

[Str] The build page url

=head2 job

[L<Net::Jenkins::Job>] The Job object 

=head2 api

[L<Net::Jenkins>] The API object 

=head1 METHODS

=head2 details

[HashRef] Get details 

=head2 name

[Str] Job name

=head2 id

[Int] Job ID

=head2 created_at

[DateTime] created at 

=head2 console

[Str] Console Output Content

=head2 console_handle

Console FH

=head2 building

[Bool] is this build is building ?

=head2 result

[Str] Build result.

=cut


lib/Net/Jenkins/Job/QueueItem.pm
package Net::Jenkins::Job::QueueItem;
use Moose;
use methods;
use Net::Jenkins::Job;

has id  => ( is => 'rw' );

has why  => ( is => 'rw' );

has stuck  => ( is => 'rw' );

has buildable  => ( is => 'rw' );

has inQueueSince  => ( is => 'rw' );

has params  => ( is => 'rw' );

has timestamp  => ( is => 'rw' );

has blocked  => ( is => 'rw' );

has job => ( is => 'rw' );

has _api => ( is => 'rw' , isa => 'Net::Jenkins' );

sub BUILDARGS {
    my ($self,%args) = @_;
    $args{job} = Net::Jenkins::Job->new( %{ $args{task} } , $args{_api} ) if $args{task} && ! $args{job};
    return \%args;
}

=pod 

{
    'stuck' => $VAR1->{'concurrentBuild'},
    'buildable' => $VAR1->{'concurrentBuild'},
    'task' => {
                'url' => 'http://localhost:8080/job/Phifty/',
                'name' => 'Phifty'
                },
    'inQueueSince' => '1337737814477',
    'params' => '',
    'timestamp' => '1337737819477',
    'blocked' => $VAR1->{'concurrentBuild'},
    'id' => 35,
    'why' => 'In the quiet period. Expires in 4.9 sec'
},

=cut

1;


lib/Net/Jenkins.pm
package Net::Jenkins;
use strict;
use warnings;
our $VERSION = '0.06';
use Net::Jenkins::Job;
use Net::Jenkins::Job::Build;
use Net::HTTP;
use LWP::UserAgent;
use Moose;
use methods;
use URI;
use JSON;

has scheme => ( is => 'rw', isa => 'Str', default => 'http' );

has host => ( 
            is => 'rw', 
            isa => 'Str' , 
            default => sub {  
                return $ENV{JENKINS_HOST} ? $ENV{JENKINS_HOST} : 'localhost';
            });

has port => ( is => 'rw', isa => 'Int' , 
            default => sub {  
                return $ENV{JENKINS_PORT} ? $ENV{JENKINS_PORT} : 8080;
            });

has user_agent => ( is => 'rw' , default => sub { 
    return LWP::UserAgent->new;
});

method get_base_url {
    return $self->scheme 
                . '://' . $self->host 
                . ':' . $self->port;
}


method post_url ($uri,%args) {
    return $self->user_agent->post( $uri , \%args );
}

method get_url ($uri) {
    return $self->user_agent->get($uri);
}

method get_json ( $uri ) {
    my $response = $self->user_agent->get($uri);
    return decode_json $response->decoded_content if $response->is_success;
}

method summary {
    my $uri = $self->get_base_url . '/api/json';
    return $self->get_json( $uri );
}

method mode {
    return $self->summary->{mode};
}

method jobs {
    return map { Net::Jenkins::Job->new( %$_ , api => $self ) } @{ $self->summary->{jobs} };
}

method use_security {
    return $self->summary->{useSecurity};
}

method use_crumbs {
    return $self->summary->{useCrumbs};
}

method views {
    return @{ $self->summary->{views} };
}

method restart {
    my $uri = $self->get_base_url . '/restart';
    return $self->get_url( $uri )->is_success;
}


method create_job ($job_name,$xml) {
    my $uri = URI->new( $self->get_base_url . '/createItem' );
    $uri->query_form( name => $job_name );
    my $response = $self->user_agent->post( $uri , "Content-Type" => "application/xml" , Content => $xml );
    return $response->code == 200;
}

method update_job ($job_name,$xml) {
    my $uri = URI->new( $self->job_url($job_name) . '/config.xml' );
    my $response = $self->user_agent->post( $uri , "Content-Type" => "application/xml" , Content => $xml );
    return $response->code == 200;
}

method copy_job ($job_name, $from_job_name) {
    my $uri = URI->new( $self->get_base_url . '/createItem' );

    # name=NEWJOBNAME&mode=copy&from=FROMJOBNAME
    $uri->query_form( name => $job_name , from => $from_job_name , mode => 'copy' );
    my $response = $self->user_agent->post( $uri );
    return $response->code == 302 ? 1 : 0;
}

method delete_job ($job_name) {
    my $uri = $self->job_url($job_name) . '/doDelete';
    return $self->post_url( $uri )->code == 302 ? 1 : 0;
}

method disable_job ($job_name) {
    my $uri = $self->job_url($job_name) . '/disable';
    return $self->post_url( $uri )->code == 302 ? 1 : 0;
}

method enable_job ($job_name) {
    my $uri = $self->job_url($job_name) . '/enable';
    return $self->post_url( $uri )->code == 302 ? 1 : 0;
}

method build_job ($job_name) {
    my $uri = $self->job_url($job_name) . '/build';
    return $self->post_url( $uri )->code == 302 ? 1 : 0;
}


method get_job_details ($job_name) {
    return $self->get_json(
        $self->job_url($job_name) . '/api/json'
    );
}




method get_builds ($job_name) {
    my $config = $self->get_job_details($job_name);
    return @{ $config->{builds} };
}

method get_build_details ($job_name, $number) {
    my $uri = $self->job_build_url($job_name,$number) . '/api/json';
    return $self->get_json($uri);
}


# Which returns a FH::Handle
method get_build_console_handle ($job_name, $number) {
    my $uri = URI->new( $self->job_build_url($job_name,$number) . '/consoleText' );

    # http://localhost:8080/job/Phifty/1/consoleText

    my $s = Net::HTTP->new(Host => $self->host ) || die $@;
    $s->write_request(GET => $uri->path , 'User-Agent' => "Perl/Net::Jenkins");
    my($code, $mess, %h) = $s->read_response_headers;

    return $s if $code == 200;
}

method get_build_console ($job_name, $number) {
    my $s = $self->get_build_console_handle( $job_name, $number );
    return unless $s;

    my $body = '';;
    while (1) {
        my $buf;
        my $n = $s->read_entity_body($buf, 1024);
        die "build console of $job_name #$number read failed: $!" unless defined $n;
        last unless $n;
        $body .= $buf;
    }
    return $body;
}

# ================================
# URL methods
# ================================
method job_url ($job_name) {
    return $self->get_base_url . '/job/' . $job_name;
}

method job_build_url ($job_name,$number) {
    # http://localhost:8080/job/Phifty/2/api/json
    return $self->job_url($job_name) . '/' . $number;
}

1;
__END__

=head1 NAME

Net::Jenkins -

=head1 SYNOPSIS

    my $jenkins = Net::Jenkins->new;
    my $jenkins = Net::Jenkins->new( host => 'ci.machine.dev' , port => 1234 );

    my $summary = $jenkins->summary;
    my @views = $jenkins->views;
    my $mode = $jenkins->mode;

    my $xml = read_file 'xt/config.xml'; 
    if( $jenkins->create_job( 'Phifty', $xml ) ) {
        $jenkins->copy_job( 'test2' , 'Phifty' );
    }

    my @jobs = $jenkins->jobs;   # [ Net::Jenkins::Job , ... ]

    for my $job ( $jenkins->jobs ) {

        # trigger a build
        $job->build;

        my $details = $job->details;
        my $queue = $job->queue_item;

        sleep 1 while $job->in_queue ;

        if( $job->last_build ) {
            $job->last_build->console;
        }

        # Net::Jenkins::Job::Build
        for my $build ( $job->builds ) {
            my $d = $build->details;
            $build->name;
            $build->id;
            $build->created_at;  # DateTime object
        }

        $job->delete;
    }
    $jenkins->restart;  # returns true if success



=head1 DESCRIPTION

Net::Jenkins is

=head1 AUTHOR

Yo-An Lin E<lt>cornelius.howl {at} gmail.comE<gt>

=head1 SEE ALSO

L<Jenkins::Trigger>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
